src/foam/dao/index/TreeIndex.java

```diff
@@ -115,6 +116,27 @@ public Object put(Object state, FObject value) {
       prop_, key, value, tail_);
   }
 
+  @Override
+  public Object update(Object state, FObject oldValue, FObject newValue) {
+
+      Object key;
+      Object oldKey;
+      try {
+        key = prop_.f(newValue);
+        oldKey = prop_.f(oldValue);
+      } catch (ClassCastException e) {
+        return state;
+      }
+      if ( oldKey != null && key.equals(oldKey) )
+        state = ((TreeNode)state).update((TreeNode)state, prop_, key, oldValue, newValue, tail_);
+      else {
+        state =  remove(state, oldValue);
```

Comment:
I think we can make this more efficient (but maybe not easily).
If this tree contains another tree, say an index on lastname then id (a tree on primary key 'id' is always last before the ValueIndex) then while we want remove and put into the lastname tree if it was updated (say someone got married), but then this would force us to remove and put into the id tree as well, even though the id didn't change. Maybe leave this for a future PR, but at least leave a comment explaining the optimization.

Reply:
hm, I have second thoughts on this one. I believe that in case of last name change we delete id and value from "old" last name node and create "new" last name node in case there isn't any and put id and value in "new" last name node. so, id didn't change but it's now a part of "new" last name node subtree